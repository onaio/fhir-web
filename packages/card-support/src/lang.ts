import i18n, { namespace } from './mls';
import { Dictionary } from '@onaio/utils';

export type Lang = Dictionary<string>;
const lang: Lang = {};

/** recompute values */
function fill() {
  lang.ERROR_OCCURRED = i18n.t(`An error occurred`, { ns: namespace });
  lang.DOWNLOAD_CLIENT_DATA = i18n.t(`Download Client Data`, { ns: namespace });
  lang.CLIENT_LOCATION = i18n.t(`Client Location`, { ns: namespace });
  lang.ALL_LOCATIONS = i18n.t(`All Locations`, { ns: namespace });
  lang.CARD_STATUS = i18n.t(`Card Status`, { ns: namespace });
  lang.NEEDS_CARD = i18n.t(`Needs Card`, { ns: namespace });
  lang.CARD_NOT_NEEDED = i18n.t(`Card not needed`, { ns: namespace });
  lang.BOTH_NEEDS_CARD_CARD_NOT_NEEDED = i18n.t(`Both "Needs card" and "Card not needed"`, {
    ns: namespace,
  });
  lang.CARD_ORDER_DATE = i18n.t(`Card Order Date`, { ns: namespace });
  lang.CARD_ORDER_DATE_REQUIRED = i18n.t(`Please enter start date and end date`, { ns: namespace });
  lang.SELECT_CARD_ORDER_DATE = i18n.t(`Select Card Order Date`, { ns: namespace });
  lang.DOWNLOADING = i18n.t(`Downloading`, { ns: namespace });
  lang.DOWNLOAD_CSV = i18n.t(`Download CSV`, { ns: namespace });
  lang.NO_DATA_FOUND = i18n.t(`No data found`, { ns: namespace });
}

// run it initial
fill();

// bind some events and fill values again (doing the magic you expect to happen magically)
i18n.on(`languageChanged`, () => {
  fill();
});

// export the const
export default lang;
