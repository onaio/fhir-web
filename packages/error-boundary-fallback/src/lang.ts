import i18n, { namespace } from './mls';
import { Dictionary } from '@onaio/utils';

const lang: Dictionary<string> = {};

/** recompute values */
function fill() {
  lang.PAGE_TITLE = i18n.t(`An Error Occurred`, { ns: namespace });
  lang.PAGE_SUB_TITLE = i18n.t(
    `There has been an error. Itâ€™s been reported to the site administrators via email and should be fixed shortly. Thanks for your patience.'`,
    { ns: namespace }
  );
  lang.BUTTON_TITLE = i18n.t(`Back Home`, { ns: namespace });
}

// run it initial
fill();

// bind some events and fill values again (doing the magic you expect to happen magically)
i18n.on(`languageChanged`, () => {
  fill();
});

// export the const
export default lang;
