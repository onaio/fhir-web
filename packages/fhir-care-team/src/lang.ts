import i18n, { namespace } from './mls';
import { Dictionary } from '@onaio/utils';

export type Lang = Dictionary<string>;

const lang: Lang = {};

/** recompute values */
function fill() {
  // Tooltips
  lang.TIP_REQUIRED_FIELD = i18n.t(`This is a required field`, { ns: namespace });

  // Errors
  lang.ERROR_OCCURRED = i18n.t(`An error occurred`, { ns: namespace });

  // Messages
  lang.CARE_TEAMS_UPDATE_SUCCESS = i18n.t(`Successfully Updated Care Teams`, { ns: namespace });
  lang.CARE_TEAMS_ADD_SUCCESS = i18n.t(`Successfully Added Care Teams`, { ns: namespace });
  lang.CARE_TEAM_DELETE_SUCCESS = i18n.t(`Successfully Deleted Care Team`, { ns: namespace });
  lang.CONFIRM_DELETE = i18n.t(`Are you sure you want to delete this Care Team?`, {
    ns: namespace,
  });
  lang.YES = i18n.t(`Yes`, { ns: namespace });
  lang.NO = i18n.t(`No`, { ns: namespace });

  // Rendered text
  lang.STATUS = i18n.t(`Status`, { ns: namespace });
  lang.NAME = i18n.t(`Name`, { ns: namespace });
  lang.SAVING = i18n.t(`Saving`, { ns: namespace });
  lang.SAVE = i18n.t(`Save`, { ns: namespace });
  lang.ACTIVE = i18n.t(`Active`, { ns: namespace });
  lang.INACTIVE = i18n.t(`Inactive`, { ns: namespace });
  lang.CANCEL = i18n.t(`Cancel`, { ns: namespace });
  lang.EDIT_CARE_TEAM = i18n.t(`Edit Care Team`, { ns: namespace });
  lang.CREATE_CARE_TEAM = i18n.t(`Create Care Team`, { ns: namespace });
  lang.EDIT = i18n.t(`Edit`, { ns: namespace });
  lang.CREATE = i18n.t(`Create`, { ns: namespace });
  lang.IDENTIFIER = i18n.t(`Identifier`, { ns: namespace });
  lang.TEAMS = i18n.t(`Care Teams`, { ns: namespace });
  lang.CARE_TEAM_MEMBERS = i18n.t(`Care Teams`, { ns: namespace });
  lang.SEARCH = i18n.t(`Search`, { ns: namespace });
  lang.IDENTIFIER = i18n.t(`Identifier`, { ns: namespace });
  lang.CARE_TEAM_PAGE_HEADER = i18n.t(`FHIR Care Team`, { ns: namespace });
  lang.SUBJECT = i18n.t(`Subject`, { ns: namespace });
  lang.PARTICIPANTS = i18n.t(`Participant`, { ns: namespace });
}

// run it initial
fill();

// bind some events and fill values again (doing the magic you expect to happen magically)
i18n.on(`languageChanged`, () => {
  fill();
});

// export the const
export default lang;
