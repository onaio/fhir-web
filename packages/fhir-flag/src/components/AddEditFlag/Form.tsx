import React from 'react';
import { Select, Button, Form, Input, Space, Col } from 'antd';

const { Item: FormItem } = Form;
const { TextArea } = Input;

export interface FlagFormFields {
  id?: string;
  servicePoint?: string;
  product?: boolean;
  status?: string;
  comments?: string;
}
interface FlagFormProps {
  fhirBaseUrl: string;
  initialValues: FlagFormFields;
  disabled: string[];
  cancelUrl?: string;
  successUrl?: string;
}

const defaultProps = {
  initialValues: {},
  disabled: [],
};

const FlagForm = (props: FlagFormProps) => {
  const { initialValues } = props;

  return (
    <Col className="bg-white p-3" span={24}>
      <Form initialValues={initialValues}>
        <FormItem id="service-point" label="Service Point">
          <Input disabled value="service-point" />
        </FormItem>

        <FormItem name="product" id="product" label="Product">
          <Input value="product" disabled />
        </FormItem>

        <FormItem id="type" label="Type">
          <Select
            options={[
              { label: 'Active', value: 'active' },
              { label: 'Inactive', value: 'inactive' },
            ]}
          />
        </FormItem>

        <FormItem id="type" label="Type">
          <TextArea />
        </FormItem>

        <FormItem>
          <Space>
            <Button type="primary" id="submit-button" htmlType="submit">
              Save
            </Button>
            <Button id="cancel-button">Cancel</Button>
          </Space>
        </FormItem>
      </Form>
    </Col>
  );
};

FlagForm.defaultProps = defaultProps;

export { FlagForm };
