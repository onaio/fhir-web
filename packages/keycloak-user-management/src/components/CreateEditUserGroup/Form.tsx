import React, { useState } from 'react';
import { useHistory } from 'react-router';
import { Button, Col, Row, Form, Input, Transfer } from 'antd';
import reducerRegistry from '@onaio/redux-reducer-registry';
import { KeycloakService } from '@opensrp/keycloak-service';
import { sendErrorNotification, sendSuccessNotification } from '@opensrp/notifications';
import { KEYCLOAK_URL_USER_GROUPS, URL_USER_GROUPS } from '../../constants';
import {
  CANCEL,
  SAVE,
  SAVING,
  NAME_REQUIRED,
  NAME,
  EDIT_USER_GROUP,
  MESSAGE_USER_GROUP_EDITED,
  MESSAGE_USER_GROUP_CREATED,
  ADD_USER_GROUP,
  ERROR_OCCURED,
  REALM_ROLES,
  AVAILABLE_ROLES,
  ASSIGNED_ROLES,
  EFFECTIVE_ROLES,
  LIST_IS_EMPTY,
  SEARCH,
} from '../../lang';
import { KeycloakUserGroup } from '../../ducks/userGroups';
import { assignRoles, removeAssignedRoles } from './utils';
import {
  reducerName as keycloakUserRolesReducerName,
  reducer as keycloakUserRolesReducer,
  KeycloakUserRole,
} from '../../ducks/userRoles';

/** Register reducer */
reducerRegistry.register(keycloakUserRolesReducerName, keycloakUserRolesReducer);

/** Interface for practitioner json object */
export interface Practitioner {
  active: boolean;
  identifier: string;
  name: string;
  userId: string;
  username: string;
}
/** props for editing a user view */
export interface UserGroupFormProps {
  allRoles: KeycloakUserRole[];
  assignedRoles: KeycloakUserRole[];
  availableRoles: KeycloakUserRole[];
  effectiveRoles: KeycloakUserRole[];
  initialValues: KeycloakUserGroup;
  keycloakBaseURL: string;
}
/** default form initial values */
export const defaultInitialValues: KeycloakUserGroup = {
  access: {
    view: false,
    manage: false,
    manageMembership: false,
  },
  attributes: {},
  clientRoles: {},
  id: '',
  name: '',
  path: '',
  realmRoles: [],
  subGroups: [],
};
/** default props for editing user component */
export const defaultProps: Partial<UserGroupFormProps> = {
  allRoles: [],
  assignedRoles: [],
  availableRoles: [],
  effectiveRoles: [],
  initialValues: defaultInitialValues,
  keycloakBaseURL: '',
};

/** Util function that updates assigned/available roles on keycloak
 *
 * @param {KeycloakUserGroup} initialValues - form initial values
 * @param {string[]} targetSelectedKeys - target choice box selected keys
 * @param {string[]} sourceSelectedKeys - source choice box selected keys
 * @param {string} keycloakBaseURL - keycloak api base url
 * @param {KeycloakUserRole[]} roles - list of all keycloak realm roles
 */
export const handleTransferChange = async (
  initialValues: KeycloakUserGroup,
  targetSelectedKeys: string[],
  sourceSelectedKeys: string[],
  keycloakBaseURL: string,
  roles: KeycloakUserRole[]
) => {
  if (targetSelectedKeys.length) {
    await removeAssignedRoles(initialValues.id, keycloakBaseURL, roles, targetSelectedKeys);
  } else if (sourceSelectedKeys.length) {
    await assignRoles(initialValues.id, keycloakBaseURL, roles, sourceSelectedKeys);
  }
};

/** User group form for editing/adding user groups
 *
 * @param {object} props - component props
 */
const UserGroupForm: React.FC<UserGroupFormProps> = (props: UserGroupFormProps) => {
  const {
    initialValues,
    keycloakBaseURL,
    assignedRoles,
    availableRoles,
    effectiveRoles,
    allRoles,
  } = props;
  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
  const [sourceSelectedKeys, setSourceSelectedKeys] = useState<string[]>([]);
  const [targetSelectedKeys, setTargetSelectedKeys] = useState<string[]>([]);
  const [targetKeys, setTargetKeys] = useState<string[]>([]);
  const history = useHistory();
  const [form] = Form.useForm();
  const layout = {
    labelCol: {
      xs: { offset: 0, span: 16 },
      sm: { offset: 2, span: 10 },
      md: { offset: 0, span: 8 },
      lg: { offset: 0, span: 6 },
    },
    wrapperCol: { xs: { span: 24 }, sm: { span: 14 }, md: { span: 12 }, lg: { span: 10 } },
  };
  const tailLayout = {
    wrapperCol: {
      xs: { offset: 0, span: 16 },
      sm: { offset: 12, span: 24 },
      md: { offset: 8, span: 16 },
      lg: { offset: 6, span: 14 },
    },
  };

  /** Update form initial values when initialValues prop changes, without this
   * the form fields initial values will not change if props.initiaValues is updated
   * **/
  React.useEffect(() => {
    form.setFieldsValue({
      ...initialValues,
    });
  }, [form, initialValues]);

  const onChange = async (nextTargetKeys: string[]) => {
    // only add finally block since catch has already
    // been handled in function definition
    handleTransferChange(
      initialValues,
      targetSelectedKeys,
      sourceSelectedKeys,
      keycloakBaseURL,
      allRoles
    ).finally(() => setTargetKeys(nextTargetKeys));
  };

  const onSelectChange = (sourceSelectedKeys: string[], targetSelectedKeys: string[]) => {
    setSourceSelectedKeys([...sourceSelectedKeys]);
    setTargetSelectedKeys([...targetSelectedKeys]);
  };

  const data = [...assignedRoles, ...availableRoles].map((item: KeycloakUserRole) => ({
    key: item.id,
    title: item.name,
  }));

  return (
    <Row className="layout-content user-group">
      {/** If email is provided render edit group otherwise add group */}
      <h5 className="mb-3 header-title">
        {props.initialValues.id ? `${EDIT_USER_GROUP} | ${initialValues.name}` : ADD_USER_GROUP}
      </h5>
      <Col className="bg-white p-3" span={24}>
        <Form
          {...layout}
          form={form}
          initialValues={{
            ...initialValues,
          }}
          onFinish={(values: KeycloakUserGroup & { roles?: string[] }) => {
            // remove roles array from payload
            delete values.roles;
            setIsSubmitting(true);
            if (initialValues.id) {
              const serve = new KeycloakService(
                `${KEYCLOAK_URL_USER_GROUPS}/${initialValues.id}`,
                keycloakBaseURL
              );
              serve
                .update(values)
                .then(() => sendSuccessNotification(MESSAGE_USER_GROUP_EDITED))
                .catch((_: Error) => sendErrorNotification(ERROR_OCCURED))
                .finally(() => setIsSubmitting(false));
            } else {
              const serve = new KeycloakService(KEYCLOAK_URL_USER_GROUPS, keycloakBaseURL);
              serve
                .create({ name: values.name })
                .then(() => sendSuccessNotification(MESSAGE_USER_GROUP_CREATED))
                .catch((_: Error) => sendErrorNotification(ERROR_OCCURED))
                .finally(() => setIsSubmitting(false));
            }
          }}
        >
          <Form.Item
            name="name"
            id="name"
            label={NAME}
            rules={[{ required: true, message: NAME_REQUIRED }]}
          >
            <Input />
          </Form.Item>
          {initialValues.id ? (
            <Form.Item name="roles" id="roles" label={REALM_ROLES}>
              <Transfer
                dataSource={data}
                titles={[AVAILABLE_ROLES, ASSIGNED_ROLES]}
                listStyle={{ flexGrow: 'inherit' }}
                targetKeys={
                  targetKeys.length
                    ? targetKeys
                    : assignedRoles.map((role: KeycloakUserRole) => role.id)
                }
                selectedKeys={[...sourceSelectedKeys, ...targetSelectedKeys]}
                render={(item) => <div>{item.title}</div>}
                disabled={false}
                onChange={onChange}
                onSelectChange={onSelectChange}
                locale={{
                  notFoundContent: LIST_IS_EMPTY,
                  searchPlaceholder: SEARCH,
                }}
              />
              {/** custom transfer to list effective roles */}
              <div className="ant-transfer">
                <div className="ant-transfer-list">
                  <div className="ant-transfer-list-header">
                    <span className="ant-transfer-list-header-title">{EFFECTIVE_ROLES}</span>
                  </div>
                  <div className="ant-transfer-list-body">
                    {!effectiveRoles.length ? (
                      <div className="ant-transfer-list-body-not-found">{LIST_IS_EMPTY}</div>
                    ) : (
                      <ul className="ant-transfer-list-content">
                        {effectiveRoles.map((role: KeycloakUserRole) => (
                          <li
                            key={role.id}
                            className="ant-transfer-list-content-item ant-transfer-list-content-item-disabled"
                          >
                            <label className="ant-checkbox-wrapper">
                              <span className="ant-checkbox">
                                <input type="checkbox" className="ant-checkbox-input" />
                                <span className="ant-checkbox-inner" />
                              </span>
                            </label>
                            <span className="ant-transfer-list-content-item-text">
                              <div>{role.name}</div>
                            </span>
                          </li>
                        ))}
                      </ul>
                    )}
                  </div>
                </div>
              </div>
            </Form.Item>
          ) : (
            ''
          )}
          <Form.Item {...tailLayout}>
            <Button type="primary" htmlType="submit" className="create-group">
              {isSubmitting ? SAVING : SAVE}
            </Button>
            <Button onClick={() => history.push(URL_USER_GROUPS)} className="cancel-group">
              {CANCEL}
            </Button>
          </Form.Item>
        </Form>
      </Col>
    </Row>
  );
};
UserGroupForm.defaultProps = defaultProps;
export { UserGroupForm };
