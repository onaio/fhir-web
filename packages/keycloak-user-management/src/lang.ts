export const USER_MANAGEMENT_PAGE_HEADER = 'User Management';
export const ADD_USER = 'Add User';
