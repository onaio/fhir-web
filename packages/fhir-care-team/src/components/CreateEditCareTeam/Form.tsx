import React, { useState } from 'react';
import { useHistory } from 'react-router';
import { Button, Col, Row, Form, Input, Radio, Select } from 'antd';
import { sendErrorNotification } from '@opensrp/notifications';
import lang from '../../lang';
import { submitForm } from './utils';
import { URL_CARE_TEAM } from '../../constants';

export interface Fields {
  id: string;
  name: string;
}

export interface FormFields {
  uuid: string | undefined;
  id: string | undefined;
  name: string | undefined;
  status: string | undefined;
  practitionersId?: string[];
  groupsId?: string;
}

export interface CareTeamFormProps {
  initialValues: FormFields;
  fhirBaseURL: string;
  practitioners: Fields[];
  groups: Fields[];
}

/** Care Team form for editing/adding FHIR Care Teams
 *
 * @param {object} props - component props
 */
const CareTeamForm: React.FC<CareTeamFormProps> = (props: CareTeamFormProps) => {
  const { fhirBaseURL, initialValues } = props;
  const [isSubmitting, setIsSubmitting] = useState<boolean>(false);
  const history = useHistory();
  const [form] = Form.useForm();
  const layout = {
    labelCol: {
      xs: { offset: 0, span: 16 },
      sm: { offset: 2, span: 10 },
      md: { offset: 0, span: 8 },
      lg: { offset: 0, span: 6 },
    },
    wrapperCol: { xs: { span: 24 }, sm: { span: 14 }, md: { span: 12 }, lg: { span: 10 } },
  };
  const tailLayout = {
    wrapperCol: {
      xs: { offset: 0, span: 16 },
      sm: { offset: 12, span: 24 },
      md: { offset: 8, span: 16 },
      lg: { offset: 6, span: 14 },
    },
  };

  const status = [
    { label: 'Active', value: 'active' },
    { label: 'Inactive', value: 'inactive' },
  ];

  /** Update form initial values when initialValues prop changes, without this
   * the form fields initial values will not change if props.initiaValues is updated
   * **/
  React.useEffect(() => {
    form.setFieldsValue({
      ...(initialValues as FormFields),
    });
  }, [form, initialValues]);

  return (
    <Row className="layout-content user-group">
      {/** If email is provided render edit group otherwise add group */}
      <h5 className="mb-3 header-title">
        {props.initialValues?.id
          ? `${lang.EDIT_CARE_TEAM} | ${initialValues.name}`
          : lang.CREATE_CARE_TEAM}
      </h5>
      <Col className="bg-white p-3" span={24}>
        <Form
          {...layout}
          form={form}
          initialValues={initialValues}
          onFinish={(values: FormFields) => {
            setIsSubmitting(true);
            submitForm(
              { ...initialValues, ...values },
              fhirBaseURL,
              props.initialValues?.id,
              props.initialValues?.uuid
            ).catch(() => sendErrorNotification(lang.ERROR_OCCURED));
          }}
        >
          <Form.Item id={'uuid'} hidden={true} name={'uuid'} label={'UUID'}>
            <Input />
          </Form.Item>
          <Form.Item
            name="name"
            id="name"
            label={lang.NAME}
            rules={[{ required: true, message: lang.NAME_REQUIRED }]}
          >
            <Input />
          </Form.Item>
          <Form.Item id="status" name="status" label={lang.STATUS}>
            <Radio.Group name="status">
              {status.map((e) => (
                <Radio name="status" key={e.label} value={e.value}>
                  {e.label}
                </Radio>
              ))}
            </Radio.Group>
          </Form.Item>
          <Form.Item
            name="practitionersId"
            id="practitionersId"
            label={lang.PARTICIPANTS}
            tooltip={lang.TIP_REQUIRED_FIELD}
          >
            <Select placeholder={lang.PARTICIPANTS} allowClear mode="multiple">
              {props.practitioners.map((practitioner: Fields) => (
                <Select.Option key={practitioner.id} value={practitioner.id}>
                  {practitioner.name}
                </Select.Option>
              ))}
            </Select>
          </Form.Item>
          <Form.Item
            name="groupsId"
            id="groupsId"
            label={lang.SUBJECT}
            tooltip={lang.TIP_REQUIRED_FIELD}
          >
            <Select placeholder={lang.SUBJECT}>
              {props.groups.map((group: Fields) => (
                <Select.Option key={group.id} value={group.id}>
                  {group.name}
                </Select.Option>
              ))}
            </Select>
          </Form.Item>
          <Form.Item {...tailLayout}>
            <Button type="primary" htmlType="submit" className="create-group">
              {isSubmitting ? lang.SAVING : lang.SAVE}
            </Button>
            <Button onClick={() => history.push(URL_CARE_TEAM)} className="cancel-care-team">
              {lang.CANCEL}
            </Button>
          </Form.Item>
        </Form>
      </Col>
    </Row>
  );
};

export { CareTeamForm };
