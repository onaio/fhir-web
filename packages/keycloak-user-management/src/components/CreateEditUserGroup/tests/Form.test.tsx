import { mount, shallow } from 'enzyme';
import flushPromises from 'flush-promises';
import React from 'react';
import fetch from 'jest-fetch-mock';
import { history } from '@onaio/connected-reducer-registry';
import * as notifications from '@opensrp/notifications';
import { store } from '@opensrp/store';
import { authenticateUser } from '@onaio/session-reducer';
import * as fixtures from './fixtures';
import {
  userRoles,
  assignedRoles,
  availableRoles,
  effectiveRoles,
} from '../../../ducks/tests/fixtures';
import { act } from 'react-dom/test-utils';
import { Router } from 'react-router';
import { UserGroupForm, defaultInitialValues } from '../Form';
import toJson from 'enzyme-to-json';
import lang from '../../../lang';

jest.mock('@opensrp/notifications', () => ({
  __esModule: true,
  ...Object.assign({}, jest.requireActual('@opensrp/notifications')),
}));

describe('components/forms/UserFroupForm', () => {
  const props = {
    initialValues: defaultInitialValues,
    allRoles: userRoles,
    availableRoles,
    assignedRoles,
    effectiveRoles,
    keycloakBaseURL: 'https://keycloak-stage.smartregister.org/auth/admin/realms/opensrp-web-stage',
  };

  beforeAll(() => {
    store.dispatch(
      authenticateUser(
        true,
        {
          email: 'bob@example.com',
          name: 'Bobbie',
          username: 'RobertBaratheon',
        },
        // eslint-disable-next-line @typescript-eslint/camelcase
        { api_token: 'hunter2', oAuth2Data: { access_token: 'access token', state: 'abcde' } }
      )
    );
  });

  afterEach(() => {
    fetch.resetMocks();
    jest.resetAllMocks();
  });

  it('renders without crashing', () => {
    shallow(<UserGroupForm {...props} />);
  });

  it('renders correctly', async () => {
    const wrapper = mount(<UserGroupForm {...props} />);
    expect(wrapper.find('Row').at(0).text()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('form validation works for required fields', async () => {
    const wrapper = mount(<UserGroupForm {...props} />);

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
    });
    wrapper.update();

    expect(wrapper.find('FormItemInput').at(0).prop('errors')).toEqual(['Name is required']);

    wrapper.unmount();
  });

  it('adds new user group successfully', async () => {
    const wrapper = mount(<UserGroupForm {...props} />);
    const mockNotificationSuccess = jest.spyOn(notifications, 'sendSuccessNotification');

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    const nameInput = wrapper.find('input#name');
    nameInput.simulate('change', { target: { name: 'name', value: 'Test' } });
    wrapper.update();

    wrapper.find('form').simulate('submit');
    await act(async () => {
      await flushPromises();
      wrapper.update();
    });
    expect(mockNotificationSuccess).toHaveBeenCalledWith('User Group created successfully');
    wrapper.unmount();
  });

  it('edits user group', async () => {
    const propEdit = {
      ...props,
      initialValues: fixtures.userGroup,
    };
    const wrapper = mount(<UserGroupForm {...propEdit} />);

    // usergroup name
    await act(async () => {
      const nameInput = wrapper.find('input#name');
      nameInput.simulate('change', { target: { name: 'name', value: 'Test1' } });
    });
    wrapper.update();

    wrapper.find('form').simulate('submit');

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    await new Promise<unknown>((resolve) => setImmediate(resolve));

    const payload = {
      name: 'Test1',
    };

    expect(fetch.mock.calls[0]).toEqual([
      `https://keycloak-stage.smartregister.org/auth/admin/realms/opensrp-web-stage/groups/${fixtures.userGroup.id}`,
      {
        'Cache-Control': 'no-cache',
        Pragma: 'no-cache',
        body: JSON.stringify(payload),
        headers: {
          accept: 'application/json',
          authorization: 'Bearer access token',
          'content-type': 'application/json;charset=UTF-8',
        },
        method: 'PUT',
      },
    ]);
    wrapper.unmount();
  });

  it('usergroup is not created if api is down', async () => {
    fetch.mockReject(() => Promise.reject('API is down'));
    const mockNotificationError = jest.spyOn(notifications, 'sendErrorNotification');
    const wrapper = mount(<UserGroupForm {...props} />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    // set usersgroup  name
    const nameInput = wrapper.find('input#name');
    nameInput.simulate('change', { target: { name: 'name', value: 'Test' } });

    wrapper.find('form').simulate('submit');
    await act(async () => {
      wrapper.update();
    });

    await new Promise<unknown>((resolve) => setImmediate(resolve));
    wrapper.update();
    expect(mockNotificationError).toHaveBeenCalledWith(lang.ERROR_OCCURED);
    wrapper.unmount();
  });

  it('usergroup is not edited if api is down', async () => {
    fetch.mockReject(() => Promise.reject('API is down'));
    const mockNotificationError = jest.spyOn(notifications, 'sendErrorNotification');
    const propEdit = {
      ...props,
      initialValues: fixtures.userGroup,
    };
    const wrapper = mount(<UserGroupForm {...propEdit} />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    // usergrop name
    const nameInput = wrapper.find('input#name');
    nameInput.simulate('change', { target: { name: 'name', value: 'Test2' } });

    wrapper.find('form').simulate('submit');

    await act(async () => {
      wrapper.update();
    });
    await new Promise<unknown>((resolve) => setImmediate(resolve));
    wrapper.update();
    expect(mockNotificationError).toHaveBeenCalledWith(lang.ERROR_OCCURED);
    wrapper.unmount();
  });

  it('cancel button returns user to groups list view  page', async () => {
    const wrapper = mount(
      <Router history={history}>
        <UserGroupForm {...props} />
      </Router>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    await new Promise<unknown>((resolve) => setImmediate(resolve));
    wrapper.update();
    const button = wrapper.find('button.cancel-group');
    button.simulate('click');
    expect(history.location.pathname).toEqual('/admin/users/groups');
    wrapper.unmount();
  });

  it('render correct user group name in header', async () => {
    fetch.mockReject(() => Promise.reject('API is down'));
    const propEdit = {
      ...props,
      initialValues: fixtures.userGroup,
    };
    const wrapper = mount(<UserGroupForm {...propEdit} />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.find('.mb-3.header-title').text()).toEqual(
      `Edit User Group | ${fixtures.userGroup.name}`
    );
    wrapper.unmount();
  });

  it('renders transfer component on edit', async () => {
    const propEdit = {
      ...props,
      initialValues: fixtures.userGroup,
    };
    const wrapper = mount(<UserGroupForm {...propEdit} />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(toJson(wrapper.find('Transfer'))).toBeTruthy();
    wrapper.unmount();
  });

  it('Test transfering role from source to target', async () => {
    const propEdit = {
      ...props,
      initialValues: fixtures.userGroup,
    };
    const wrapper = mount(<UserGroupForm {...propEdit} />);

    const transferComponent = wrapper.find('Transfer');
    const transferSource = transferComponent.find('TransferList').at(0);
    const transferTarget = transferComponent.find('TransferList').at(1);

    // check length of source and target choice boxes
    expect(toJson(transferSource.find('.ant-transfer-list-content-item'))).toHaveLength(2);
    expect(toJson(transferTarget.find('.ant-transfer-list-content-item'))).toHaveLength(6);

    // check labels for source transfer box
    expect(transferSource.find('.ant-transfer-list-content-item').at(0).text()).toEqual(
      'EDIT_KEYCLOAK_USERS'
    );
    expect(transferSource.find('.ant-transfer-list-content-item').at(1).text()).toEqual(
      'VIEW_KEYCLOAK_USERS'
    );

    // check labels for target transfer box
    expect(transferTarget.find('.ant-transfer-list-content-item').at(0).text()).toEqual('OPENMRS');
    expect(transferTarget.find('.ant-transfer-list-content-item').at(1).text()).toEqual(
      'ALL_EVENTS'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(2).text()).toEqual(
      'PLANS_FOR_USER'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(3).text()).toEqual(
      'realm-admin'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(4).text()).toEqual(
      'offline_access'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(5).text()).toEqual(
      'uma_authorization'
    );

    // trigger change on source transfer box
    wrapper
      .find('.ant-checkbox-input')
      .at(0)
      .simulate('change', { target: { checked: true } });
    wrapper.update();
    // move role to target transfer box
    expect(toJson(wrapper.find('.ant-transfer-operation button'))).toHaveLength(2);
    wrapper.find('.ant-transfer-operation button').at(0).simulate('click');

    await act(async () => {
      await flushPromises();
    });

    wrapper.update();
    // const source = wrapper.find('Transfer').find('TransferList').at(0);
    // const target = wrapper.find('Transfer').find('TransferList').at(1);

    // confirm that the roles have been transfered to target box
    expect(
      wrapper.find('Transfer').find('TransferList').at(0).find('.ant-transfer-list-content-item')
    ).toHaveLength(0);
    // check length of target choice box
    expect(
      wrapper.find('Transfer').find('TransferList').at(1).find('.ant-transfer-list-content-item')
    ).toHaveLength(8);
    wrapper.unmount();
  });

  it('Test trasnfering role from target to source', async () => {
    const propEdit = {
      ...props,
      initialValues: fixtures.userGroup,
    };
    const wrapper = mount(<UserGroupForm {...propEdit} />);

    const transferComponent = wrapper.find('Transfer');
    const transferSource = transferComponent.find('TransferList').at(0);
    const transferTarget = transferComponent.find('TransferList').at(1);

    // check length of source and target choice boxes
    expect(toJson(transferSource.find('.ant-transfer-list-content-item'))).toHaveLength(2);
    expect(toJson(transferTarget.find('.ant-transfer-list-content-item'))).toHaveLength(6);

    // check labels for source transfer box
    expect(transferSource.find('.ant-transfer-list-content-item').at(0).text()).toEqual(
      'EDIT_KEYCLOAK_USERS'
    );
    expect(transferSource.find('.ant-transfer-list-content-item').at(1).text()).toEqual(
      'VIEW_KEYCLOAK_USERS'
    );

    // check labels for target transfer box
    expect(transferTarget.find('.ant-transfer-list-content-item').at(0).text()).toEqual('OPENMRS');
    expect(transferTarget.find('.ant-transfer-list-content-item').at(1).text()).toEqual(
      'ALL_EVENTS'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(2).text()).toEqual(
      'PLANS_FOR_USER'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(3).text()).toEqual(
      'realm-admin'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(4).text()).toEqual(
      'offline_access'
    );
    expect(transferTarget.find('.ant-transfer-list-content-item').at(5).text()).toEqual(
      'uma_authorization'
    );

    // trigger change on targer transfer box
    wrapper
      .find('.ant-checkbox-input')
      .at(3)
      .simulate('change', { target: { checked: true } });
    wrapper.update();
    // move role to target transfer box
    wrapper.find('.ant-transfer-operation .ant-btn').at(1).simulate('click');

    await act(async () => {
      await flushPromises();
    });

    wrapper.update();
    expect(
      toJson(
        wrapper.find('Transfer').find('TransferList').at(0).find('.ant-transfer-list-content-item')
      )
    ).toHaveLength(2);
    // check length of target choice box
    expect(
      wrapper.find('Transfer').find('TransferList').at(1).find('.ant-transfer-list-content-item')
    ).toHaveLength(6);
    wrapper.unmount();
  });

  it('doesnt show transfer component on new user group creation', async () => {
    const wrapper = mount(<UserGroupForm {...props} />);

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(toJson(wrapper.find('Transfer'))).toBeFalsy();
    wrapper.unmount();
  });
});
