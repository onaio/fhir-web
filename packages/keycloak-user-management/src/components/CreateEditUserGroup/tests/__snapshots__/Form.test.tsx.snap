// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/forms/UserFroupForm renders correctly 1`] = `"New User GroupNameSaveCancel"`;
